// ravendb/dist/Documents/Queries/IndexQuery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Exceptions_1 = require("../../Exceptions");
const IndexQueryWithParameters_1 = require("./IndexQueryWithParameters");
const QueryHashCalculator_1 = require("./QueryHashCalculator");
const TypeUtil_1 = require("../../Utility/TypeUtil");
const Serializer_1 = require("../../Mapping/Json/Serializer");
class IndexQuery extends IndexQueryWithParameters_1.IndexQueryWithParameters {
    constructor(query) {
        super();
        this.query = query;
    }
    getQueryHash() {
        const hasher = new QueryHashCalculator_1.QueryHashCalculator();
        try {
            hasher.write(this.query);
            hasher.write(this.waitForNonStaleResults);
            hasher.write(this.skipDuplicateChecking);
            hasher.write(this.waitForNonStaleResultsTimeout || 0);
            hasher.write(this.start);
            hasher.write(this.pageSize);
            hasher.write(this.queryParameters);
            return hasher.getHash();
        }
        catch (err) {
            Exceptions_1.throwError("RavenException", "Unable to calculate hash", err);
        }
    }
}
exports.IndexQuery = IndexQuery;
function writeIndexQuery(conventions, indexQuery) {
    const result = {
        Query: indexQuery.query
    };
    if (indexQuery.pageSizeSet && indexQuery.pageSize >= 0) {
        result["PageSize"] = indexQuery.pageSize;
    }
    if (indexQuery.waitForNonStaleResults) {
        result["WaitForNonStaleResults"] = indexQuery.waitForNonStaleResults;
    }
    if (indexQuery.start > 0) {
        result["Start"] = indexQuery.start;
    }
    if (!TypeUtil_1.TypeUtil.isNullOrUndefined(indexQuery.waitForNonStaleResultsTimeout)) {
        result["WaitForNonStaleResultsTimeout"] = indexQuery.waitForNonStaleResultsTimeout;
    }
    if (indexQuery.disableCaching) {
        result["DisableCaching"] = indexQuery.disableCaching;
    }
    if (indexQuery.skipDuplicateChecking) {
        result["SkipDuplicateChecking"] = indexQuery.skipDuplicateChecking;
    }
    if (!indexQuery.queryParameters) {
        result["QueryParameters"] = null;
    }
    else {
        result["QueryParameters"] = indexQuery.queryParameters;
    }
    return Serializer_1.JsonSerializer.getDefault().serialize(result);
}
exports.writeIndexQuery = writeIndexQuery;
