// ravendb/dist/Documents/Operations/DeleteByQueryOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IndexQuery_1 = require("../Queries/IndexQuery");
const Exceptions_1 = require("../../Exceptions");
const RavenCommand_1 = require("../../Http/RavenCommand");
const StringBuilder = require("string-builder");
const TypeUtil_1 = require("../../Utility/TypeUtil");
class DeleteByQueryOperation {
    get resultType() {
        return "OperationId";
    }
    constructor(queryToDelete, options) {
        if (!queryToDelete) {
            Exceptions_1.throwError("InvalidArgumentException", "QueryToDelete cannot be null");
        }
        this._queryToDelete = queryToDelete;
        this._options = options;
    }
    getCommand(store, conventions, cache) {
        return new DeleteByIndexCommand(conventions, this._queryToDelete, this._options);
    }
}
exports.DeleteByQueryOperation = DeleteByQueryOperation;
class DeleteByIndexCommand extends RavenCommand_1.RavenCommand {
    constructor(conventions, queryToDelete, options) {
        super();
        this._conventions = conventions;
        this._queryToDelete = queryToDelete;
        this._options = options || {};
    }
    createRequest(node) {
        const path = new StringBuilder(node.url)
            .append("/databases/")
            .append(node.database)
            .append("/queries")
            .append("?allowStale=")
            .append(this._options.allowStale || "");
        if (!TypeUtil_1.TypeUtil.isNullOrUndefined(this._options.maxOpsPerSecond)) {
            path.append("&maxOpsPerSec=")
                .append(this._options.maxOpsPerSecond);
        }
        path
            .append("&details=")
            .append(this._options.retrieveDetails || "");
        if (this._options.staleTimeout) {
            path.append("&staleTimeout=")
                .append(this._options.staleTimeout);
        }
        const body = IndexQuery_1.writeIndexQuery(this._conventions, this._queryToDelete);
        const headers = this._getHeaders().withContentTypeJson().build();
        const uri = path.toString();
        return {
            uri,
            body,
            method: "DELETE",
            headers
        };
    }
    setResponse(response, fromCache) {
        if (!response) {
            this._throwInvalidResponse();
        }
        this.result = this._serializer.deserialize(response);
    }
    get isReadRequest() {
        return false;
    }
}
exports.DeleteByIndexCommand = DeleteByIndexCommand;
