// ravendb/dist/Documents/Operations/GetCollectionStatisticsOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../Http/RavenCommand");
const ObjectMapper_1 = require("../../Mapping/ObjectMapper");
const Serializer_1 = require("../../Mapping/Json/Serializer");
class GetCollectionStatisticsOperation {
    getCommand(conventions) {
        return new GetCollectionStatisticsCommand();
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.GetCollectionStatisticsOperation = GetCollectionStatisticsOperation;
class GetCollectionStatisticsCommand extends RavenCommand_1.RavenCommand {
    constructor() {
        super();
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/collections/stats";
        return { uri };
    }
    get _serializer() {
        return Serializer_1.JsonSerializer.getDefault();
    }
    setResponse(response, fromCache) {
        if (!response) {
            this._throwInvalidResponse();
        }
        const rawResult = this._serializer.deserialize(response);
        this.result = ObjectMapper_1.ObjectKeysTransform.camelCase(rawResult);
    }
}
exports.GetCollectionStatisticsCommand = GetCollectionStatisticsCommand;
