// ravendb/dist/Documents/Operations/GetOperationStateOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("../..");
const RavenCommand_1 = require("../../Http/RavenCommand");
class GetOperationStateOperation {
    constructor(id) {
        this._id = id;
    }
    getCommand(conventions) {
        return new GetOperationStateCommand(__1.DocumentConventions.defaultConventions, this._id);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.GetOperationStateOperation = GetOperationStateOperation;
class GetOperationStateCommand extends RavenCommand_1.RavenCommand {
    get isReadRequest() {
        return true;
    }
    constructor(conventions, id) {
        super();
        this._conventions = conventions;
        this._id = id;
    }
    createRequest(node) {
        const uri = `${node.url}/databases/${node.database}/operations/state?id=${this._id}`;
        return { uri };
    }
    setResponse(response, fromCache) {
        if (!response) {
            return;
        }
        this.result = this._serializer.deserialize(response);
    }
}
exports.GetOperationStateCommand = GetOperationStateCommand;
