// ravendb/dist/Documents/Operations/GetStatisticsOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../Http/RavenCommand");
class GetStatisticsOperation {
    get resultType() {
        return "CommandResult";
    }
    constructor(debugTag) {
        this._debugTag = debugTag;
    }
    getCommand(conventions) {
        return new GetStatisticsCommand(this._debugTag);
    }
}
exports.GetStatisticsOperation = GetStatisticsOperation;
class GetStatisticsCommand extends RavenCommand_1.RavenCommand {
    constructor(debugTag) {
        super();
        this._debugTag = debugTag;
    }
    createRequest(node) {
        let uri = `${node.url}/databases/${node.database}/stats`;
        if (this._debugTag) {
            uri += "?" + this._debugTag;
        }
        return { uri };
    }
    setResponse(response, fromCache) {
        this.result = this._serializer.deserialize(response);
    }
    get isReadRequest() {
        return true;
    }
}
exports.GetStatisticsCommand = GetStatisticsCommand;
