// ravendb/dist/Documents/Operations/PatchOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../Http/RavenCommand");
const Exceptions_1 = require("../../Exceptions");
const Serializer_1 = require("../../Mapping/Json/Serializer");
const ObjectMapper_1 = require("../../Mapping/ObjectMapper");
class PatchOperationResult {
}
exports.PatchOperationResult = PatchOperationResult;
class PatchOperation {
    get resultType() {
        return "PatchResult";
    }
    constructor(id, changeVector, patch, patchIfMissing = null, skipPatchIfChangeVectorMismatch = false) {
        if (!patch) {
            Exceptions_1.throwError("InvalidArgumentException", "Patch cannot be null");
        }
        if (!patch.script || !patch.script.trim()) {
            Exceptions_1.throwError("InvalidArgumentException", "Patch script cannot be null");
        }
        if (patchIfMissing && !patchIfMissing.script.trim()) {
            Exceptions_1.throwError("InvalidArgumentException", "PatchIfMissing script cannot be null");
        }
        this._id = id;
        this._changeVector = changeVector;
        this._patch = patch;
        this._patchIfMissing = patchIfMissing;
        this._skipPatchIfChangeVectorMismatch = skipPatchIfChangeVectorMismatch;
    }
    getCommand(store, conventions, cache) {
        return new PatchCommand(conventions, this._id, this._changeVector, this._patch, this._patchIfMissing, this._skipPatchIfChangeVectorMismatch, false, false);
    }
}
exports.PatchOperation = PatchOperation;
class PatchCommand extends RavenCommand_1.RavenCommand {
    constructor(conventions, id, changeVector, patch, patchIfMissing, skipPatchIfChangeVectorMismatch, returnDebugInformation, test) {
        super();
        if (!conventions) {
            Exceptions_1.throwError("InvalidArgumentException", "Conventions cannot be null");
        }
        if (!patch) {
            Exceptions_1.throwError("InvalidArgumentException", "Patch cannot be null");
        }
        if (!patch.script.trim()) {
            Exceptions_1.throwError("InvalidArgumentException", "Patch.Script cannot be null");
        }
        if (patchIfMissing && !patchIfMissing.script.trim()) {
            Exceptions_1.throwError("InvalidArgumentException", "PatchIfMissing.Script cannot be null");
        }
        if (!id) {
            Exceptions_1.throwError("InvalidArgumentException", "Id cannot be null");
        }
        this._id = id;
        this._changeVector = changeVector;
        this._patch = this._typedObjectMapper.toObjectLiteral({ patch, patchIfMissing });
        this._skipPatchIfChangeVectorMismatch = skipPatchIfChangeVectorMismatch;
        this._returnDebugInformation = returnDebugInformation;
        this._test = test;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        let uri = node.url + "/databases/" + node.database + "/docs?id=" + encodeURIComponent(this._id);
        if (this._skipPatchIfChangeVectorMismatch) {
            uri += "&skipPatchIfChangeVectorMismatch=true";
        }
        if (this._returnDebugInformation) {
            uri += "&debug=true";
        }
        if (this._test) {
            uri += "&test=true";
        }
        const body = this._serializer.serialize(this._patch);
        const req = {
            method: "PATCH",
            uri,
            headers: this._getHeaders().withContentTypeJson().build(),
            body
        };
        this._addChangeVectorIfNotNull(this._changeVector, req);
        return req;
    }
    setResponse(response, fromCache) {
        if (!response) {
            return;
        }
        const rawResult = Serializer_1.JsonSerializer.getDefault().deserialize(response);
        this.result = ObjectMapper_1.ObjectKeysTransform.camelCase(rawResult);
    }
}
exports.PatchCommand = PatchCommand;
