// ravendb/dist/TestDriver/RavenServerLocator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Exceptions_1 = require("../Exceptions");
class RavenServerLocator {
    getServerPath() {
        const serverPath = process.env[RavenServerLocator.ENV_SERVER_PATH];
        if (!serverPath) {
            Exceptions_1.throwError("InvalidOperationException", "Unable to find RavenDB server path. Please make sure "
                + RavenServerLocator.ENV_SERVER_PATH
                + " environment variable is set and is valid " +
                "(current value = " + serverPath + ")");
        }
        return serverPath;
    }
    withHttps() {
        return false;
    }
    getServerHost() {
        return "127.0.0.1";
    }
    getCommand() {
        return this.getServerPath();
    }
    getCommandArguments() {
        return [];
    }
    getServerCertificatePath() {
        return Exceptions_1.throwError("NotSupportedException");
    }
    getClientAuthOptions() {
        return Exceptions_1.throwError("NotSupportedException");
    }
}
RavenServerLocator.ENV_SERVER_PATH = "RAVENDB_TEST_SERVER_PATH";
exports.RavenServerLocator = RavenServerLocator;
