// ravendb/dist/Documents/Commands/NextIdentityForCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../Http/RavenCommand");
const Exceptions_1 = require("../../Exceptions");
class NextIdentityForCommand extends RavenCommand_1.RavenCommand {
    constructor(id) {
        super();
        if (!id) {
            Exceptions_1.throwError("InvalidArgumentException", "Id cannot be null");
        }
        this._id = id;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        RavenCommand_1.RavenCommand.ensureIsNotNullOrEmpty(this._id, "id");
        const uri = node.url + "/databases/" + node.database + "/identity/next?name=" + encodeURIComponent(this._id);
        return {
            method: "POST",
            uri
        };
    }
    setResponse(response, fromCache) {
        if (!response) {
            this._throwInvalidResponse();
        }
        const jsonNode = this._serializer.deserialize(response);
        if (!jsonNode["newIdentityValue"]) {
            this._throwInvalidResponse();
        }
        this.result = jsonNode["newIdentityValue"];
    }
}
exports.NextIdentityForCommand = NextIdentityForCommand;
