// ravendb/dist/Documents/Operations/CompareExchange/GetCompareExchangeValueOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../../Http/RavenCommand");
const Exceptions_1 = require("../../../Exceptions");
const CompareExchangeValueResultParser_1 = require("./CompareExchangeValueResultParser");
class GetCompareExchangeValueOperation {
    constructor(key, clazz) {
        this._key = key;
        this._clazz = clazz;
    }
    getCommand(store, conventions, cache) {
        return new GetCompareExchangeValueCommand(this._key, conventions, this._clazz);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.GetCompareExchangeValueOperation = GetCompareExchangeValueOperation;
class GetCompareExchangeValueCommand extends RavenCommand_1.RavenCommand {
    constructor(key, conventions, clazz) {
        super();
        if (!key) {
            Exceptions_1.throwError("InvalidArgumentException", "The key argument must have value");
        }
        this._key = key;
        this._clazz = clazz;
        this._conventions = conventions;
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/cmpxchg?key=" + encodeURIComponent(this._key);
        return { uri };
    }
    setResponse(response, fromCache) {
        this.result = CompareExchangeValueResultParser_1.CompareExchangeValueResultParser.getValue(response, this._conventions, this._clazz);
    }
}
exports.GetCompareExchangeValueCommand = GetCompareExchangeValueCommand;
