// ravendb/dist/ServerWide/Commands/GetTcpInfoCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../Http/RavenCommand");
class TcpConnectionInfo {
}
exports.TcpConnectionInfo = TcpConnectionInfo;
class GetTcpInfoCommand extends RavenCommand_1.RavenCommand {
    constructor(tag, dbName) {
        super();
        this._tag = tag;
        this._dbName = dbName;
    }
    createRequest(node) {
        let uri;
        if (!this._dbName) {
            uri = `${node.url}/info/tcp?tcp=${this._tag}`;
        }
        else {
            uri = `${node.url}/databases/${this._dbName}/info/tcp?tcp=${this._tag}`;
        }
        this.requestedNode = node;
        return {
            uri,
            method: "GET"
        };
    }
    setResponse(response, fromCache) {
        if (!response) {
            this._throwInvalidResponse();
        }
        this.result = this._serializer.deserialize(response);
    }
    get isReadRequest() {
        return true;
    }
}
exports.GetTcpInfoCommand = GetTcpInfoCommand;
