// ravendb/dist/Documents/Operations/Indexes/DeleteIndexOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
class DeleteIndexOperation {
    constructor(indexName) {
        if (!indexName) {
            Exceptions_1.throwError("InvalidArgumentException", "Index name cannot be null.");
        }
        this._indexName = indexName;
    }
    getCommand(conventions) {
        return new DeleteIndexCommand(this._indexName);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.DeleteIndexOperation = DeleteIndexOperation;
class DeleteIndexCommand extends RavenCommand_1.RavenCommand {
    constructor(indexName) {
        super();
        this._responseType = "Empty";
        if (!indexName) {
            Exceptions_1.throwError("InvalidArgumentException", "Index name cannot be null.");
        }
        this._indexName = indexName;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database
            + "/indexes?name=" + encodeURIComponent(this._indexName);
        return { method: "DELETE", uri };
    }
    get isReadRequest() {
        return false;
    }
}
exports.DeleteIndexCommand = DeleteIndexCommand;
