// ravendb/dist/Documents/Operations/Indexes/DisableIndexOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
class DisableIndexOperation {
    constructor(indexName) {
        if (!indexName) {
            Exceptions_1.throwError("InvalidArgumentException", "IndexName cannot be null");
        }
        this._indexName = indexName;
    }
    getCommand(conventions) {
        return new DisableIndexCommand(this._indexName);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.DisableIndexOperation = DisableIndexOperation;
class DisableIndexCommand extends RavenCommand_1.RavenCommand {
    get isReadRequest() {
        return false;
    }
    constructor(indexName) {
        super();
        if (!indexName) {
            Exceptions_1.throwError("InvalidArgumentException", "IndexName cannot be null");
        }
        this._indexName = indexName;
    }
    createRequest(node) {
        const uri = node.url + "/databases/"
            + node.database + "/admin/indexes/disable?name=" + encodeURIComponent(this._indexName);
        return { method: "POST", uri };
    }
}
exports.DisableIndexCommand = DisableIndexCommand;
