// ravendb/dist/Documents/Operations/Indexes/GetIndexStatisticsOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IndexStats_1 = require("../../Indexes/IndexStats");
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
class GetIndexStatisticsOperation {
    constructor(indexName) {
        if (!indexName) {
            Exceptions_1.throwError("InvalidArgumentException", "IndexName cannot be null.");
        }
        this._indexName = indexName;
    }
    getCommand(conventions) {
        return new GetIndexStatisticsCommand(this._indexName);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.GetIndexStatisticsOperation = GetIndexStatisticsOperation;
class GetIndexStatisticsCommand extends RavenCommand_1.RavenCommand {
    constructor(indexName) {
        super();
        if (!indexName) {
            Exceptions_1.throwError("InvalidArgumentException", "IndexName cannot be null.");
        }
        this._indexName = indexName;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database
            + "/indexes/stats?name=" + encodeURIComponent(this._indexName);
        return { uri };
    }
    setResponse(response, fromCache) {
        if (!response) {
            this._throwInvalidResponse();
        }
        const responseObj = this._parseResponseDefault(response, {
            nestedTypes: {
                "results[].collections": "Map",
                "results[].collections$MAP": "CollectionStats"
            }
        }, new Map([[IndexStats_1.CollectionStats.name, IndexStats_1.CollectionStats]]));
        const results = responseObj["results"];
        if (!results.length) {
            this._throwInvalidResponse();
        }
        this.result = results[0];
    }
    get isReadRequest() {
        return true;
    }
}
exports.GetIndexStatisticsCommand = GetIndexStatisticsCommand;
