// ravendb/dist/Documents/Operations/Indexes/IndexHasChangedOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
const HttpUtil_1 = require("../../../Utility/HttpUtil");
class IndexHasChangedOperation {
    constructor(definition) {
        if (!definition) {
            Exceptions_1.throwError("InvalidArgumentException", "IndexDefinition cannot be null");
        }
        this._definition = definition;
    }
    getCommand(conventions) {
        return new IndexHasChangedCommand(conventions, this._definition);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.IndexHasChangedOperation = IndexHasChangedOperation;
class IndexHasChangedCommand extends RavenCommand_1.RavenCommand {
    constructor(conventions, definition) {
        super();
        this._definition = this._typedObjectMapper.toObjectLiteral(definition);
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/indexes/has-changed";
        const body = this._serializer.serialize(this._definition);
        const headers = HttpUtil_1.HeadersBuilder.create()
            .withContentTypeJson().build();
        return {
            method: "POST",
            uri,
            body,
            headers
        };
    }
    setResponse(response, fromCache) {
        if (!response) {
            this._throwInvalidResponse();
        }
        const resObj = this._serializer.deserialize(response);
        this.result = resObj["changed"];
    }
}
exports.IndexHasChangedCommand = IndexHasChangedCommand;
