// ravendb/dist/Documents/Operations/Indexes/SetIndexesLockOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../../Http/RavenCommand");
const Exceptions_1 = require("../../../Exceptions");
const TypeUtil_1 = require("../../../Utility/TypeUtil");
class SetIndexesLockOperation {
    constructor(paramsOrIndexName, mode) {
        if (TypeUtil_1.TypeUtil.isString(paramsOrIndexName)) {
            const indexName = paramsOrIndexName;
            if (!indexName) {
                Exceptions_1.throwError("InvalidArgumentException", "IndexName cannot be null.");
            }
            this._parameters = {
                indexNames: [indexName],
                mode
            };
        }
        else {
            const parameters = paramsOrIndexName;
            if (!parameters) {
                Exceptions_1.throwError("InvalidArgumentException", "Parameters cannot be null.");
            }
            if (!parameters.indexNames || !parameters.indexNames.length) {
                Exceptions_1.throwError("InvalidArgumentException", "IndexNames cannot be null or empty.");
            }
            this._parameters = parameters;
        }
    }
    getCommand(conventions) {
        return new SetIndexLockCommand(conventions, this._parameters);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.SetIndexesLockOperation = SetIndexesLockOperation;
class SetIndexLockCommand extends RavenCommand_1.RavenCommand {
    constructor(conventions, parameters) {
        super();
        if (!conventions) {
            Exceptions_1.throwError("InvalidArgumentException", "Conventions cannot be null");
        }
        if (!parameters) {
            Exceptions_1.throwError("InvalidArgumentException", "Parameters cannot be null");
        }
        this._parameters = this._typedObjectMapper.toObjectLiteral(parameters);
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/indexes/set-lock";
        const body = this._serializer.serialize(this._parameters);
        const headers = this._getHeaders()
            .withContentTypeJson().build();
        return {
            uri,
            method: "POST",
            body,
            headers
        };
    }
    get isReadRequest() {
        return false;
    }
}
exports.SetIndexLockCommand = SetIndexLockCommand;
