// ravendb/dist/ServerWide/Operations/GetDatabaseNamesOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../Http/RavenCommand");
class GetDatabaseNamesOperation {
    constructor(start, pageSize) {
        this._start = start;
        this._pageSize = pageSize;
    }
    getCommand(conventions) {
        return new GetDatabaseNamesCommand(this._start, this._pageSize);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.GetDatabaseNamesOperation = GetDatabaseNamesOperation;
class GetDatabaseNamesCommand extends RavenCommand_1.RavenCommand {
    constructor(start, pageSize) {
        super();
        this._start = start;
        this._pageSize = pageSize;
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = `${node.url}/databases?start=${this._start}&pageSize=${this._pageSize}&namesOnly=true`;
        return { uri };
    }
    setResponse(response, fromCache) {
        if (!response) {
            this._throwInvalidResponse();
            return;
        }
        const { databases } = this._serializer.deserialize(response);
        if (!databases || !Array.isArray(databases) || !databases.length) {
            this._throwInvalidResponse();
        }
        this.result = databases;
    }
}
exports.GetDatabaseNamesCommand = GetDatabaseNamesCommand;
