// ravendb/dist/ServerWide/Operations/GetServerWideOperationStateOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../Http/RavenCommand");
const __1 = require("../..");
class GetServerWideOperationStateOperation {
    get resultType() {
        return "CommandResult";
    }
    constructor(id) {
        this._id = id;
    }
    getCommand(conventions) {
        return new GetServerWideOperationStateCommand(__1.DocumentConventions.defaultConventions, this._id);
    }
}
exports.GetServerWideOperationStateOperation = GetServerWideOperationStateOperation;
class GetServerWideOperationStateCommand extends RavenCommand_1.RavenCommand {
    constructor(conventions, id) {
        super();
        this._conventions = conventions;
        this._id = id;
    }
    createRequest(node) {
        const uri = node.url + "/operations/state?id=" + this._id;
        return {
            uri,
            method: "GET"
        };
    }
    get isReadRequest() {
        return true;
    }
    setResponse(response, fromCache) {
        if (!response) {
            return;
        }
        this.result = this._serializer.deserialize(response);
    }
}
exports.GetServerWideOperationStateCommand = GetServerWideOperationStateCommand;
