// ravendb/dist/Documents/Operations/Configuration/PutClientConfigurationOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Exceptions_1 = require("../../../Exceptions");
const RavenCommand_1 = require("../../../Http/RavenCommand");
class PutClientConfigurationOperation {
    get resultType() {
        return "CommandResult";
    }
    constructor(configuration) {
        if (!configuration) {
            Exceptions_1.throwError("InvalidArgumentException", "Configuration cannot be null or undefined.");
        }
        this.configuration = configuration;
    }
    getCommand(conventions) {
        return new PutClientConfigurationCommand(conventions, this.configuration);
    }
}
exports.PutClientConfigurationOperation = PutClientConfigurationOperation;
class PutClientConfigurationCommand extends RavenCommand_1.RavenCommand {
    get isReadRequest() {
        return false;
    }
    get responseType() {
        return "Empty";
    }
    constructor(conventions, configuration) {
        super();
        if (!conventions) {
            Exceptions_1.throwError("InvalidArgumentException", "Document conventions cannot be null or undefined.");
        }
        if (!configuration) {
            Exceptions_1.throwError("InvalidArgumentException", "Configuration cannot be null or undefined.");
        }
        this._configuration = this._serializer.serialize(configuration);
    }
    createRequest(node) {
        const uri = `${node.url}/databases/${node.database}/admin/configuration/client`;
        return {
            method: "PUT",
            uri,
            body: this._configuration,
            headers: {
                "content-type": "application/json"
            }
        };
    }
}
exports.PutClientConfigurationCommand = PutClientConfigurationCommand;
