// ravendb/dist/Documents/Operations/ConnectionStrings/PutConnectionStringOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RavenCommand_1 = require("../../../Http/RavenCommand");
class PutConnectionStringOperation {
    constructor(connectionString) {
        this._connectionString = connectionString;
    }
    getCommand(conventions) {
        return new PutConnectionStringCommand(this._connectionString);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.PutConnectionStringOperation = PutConnectionStringOperation;
class PutConnectionStringCommand extends RavenCommand_1.RavenCommand {
    constructor(connectionString) {
        super();
        this._connectionString = connectionString;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/connection-strings";
        const headers = this._getHeaders()
            .withContentTypeJson()
            .build();
        const body = this._serializer.serialize(this._connectionString);
        return {
            method: "PUT",
            uri,
            headers,
            body
        };
    }
    setResponse(response, fromCache) {
        if (!response) {
            this._throwInvalidResponse();
        }
        this.result = this._serializer.deserialize(response);
    }
}
exports.PutConnectionStringCommand = PutConnectionStringCommand;
